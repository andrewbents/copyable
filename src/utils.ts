/**
 * Returns constructor function of an object instance.
 * Example:
 *   class User { }
 *
 *   const user = new User();
 *   getConstructor(user); // returns User
 */
export const getConstructor = <T extends Object>(obj: T) =>
  Object.getPrototypeOf(obj).constructor;
