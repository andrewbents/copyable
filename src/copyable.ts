import { copyablePrinter } from './printer';
import { getConstructor } from './utils';

type NonFunctionPropertyNames<T> = {
  [K in keyof T]: T[K] extends Function ? never : K
}[keyof T];
type NonFunctionProperties<T> = Pick<T, NonFunctionPropertyNames<T>>;

type Patch<T> = Partial<NonFunctionProperties<T>>;
type PatchKey<T> = keyof Patch<T>;

type Mapper<V> = (val: V) => V;
type MapPatch<T> = { [K in PatchKey<T>]?: Mapper<T[K]> };

export class Copyable {
  /**
   * Returns a copy of the object.
   * Has two overloads that accept
   *   1. An object that contains new values for specified properties:
   *
   *     class Person extends Copyable {
   *       constructor(readonly name: string, readonly age: number) {
   *         super()
   *       }
   *     }
   *
   *     const person1 = new Person('John', 25);
   *     person1.copy({ name: 'James', age: 20 });
   *     // -> Person(name="James", age=20)
   *
   *   2. A key and a new value for it:
   *
   *     class Book extends Copyable {
   *       constructor(readonly title: string) {
   *         super()
   *       }
   *     }
   *
   *     const book1 = new Book('TypeScript Deep Dive');
   *     book1.copy('title', 'TypeScript handbook');
   *     // -> Book(title="TypeScript handbook")
   */
  copy(patch?: Patch<this>): this;
  copy<K extends PatchKey<this>>(key: K, val: this[K]): this;
  copy(patchOrKey?: any, val?: any): this {
    if (arguments.length < 2) {
      const constructor = getConstructor(this);
      return Object.assign(new constructor(), this, patchOrKey);
    }

    const copy = this.copy();
    copy[patchOrKey] = val;
    return copy;
  }

  /**
   * Maps the object.
   * Allows to copy the object based on its current values.
   * Has two overloads that accept
   *   1. An object that contains mapper for specified properties:
   *
   *     const person1 = new Person('John', 25);
   *     // John got older
   *     const person2 = person1.map({
   *       name: name => 'Older ' + name,
   *       age: age => age + 20
   *     });
   *     // -> Person(name="Older John", age=45)
   *
   *   2. A key and a mapper for it:
   *
   *     const book1 = new Book('TypeScript Deep Dive');
   *     book1.map('title', title => title + ' Part 2');
   *     // -> Book(title="TypeScript Deep Dive Part 2")
   */
  map(patchMap?: MapPatch<this>): this;
  map<K extends PatchKey<this>>(key: K, mapper: Mapper<this[K]>): this;
  map(patchMapOrKey: any, mapper?: any): this {
    const copy = this.copy();

    if (arguments.length === 2) {
      const key = patchMapOrKey;
      copy[key] = mapper(copy[key]);
    } else {
      const patchMap = patchMapOrKey;
      for (const key in patchMap) {
        if (patchMap.hasOwnProperty(key)) {
          copy[key] = patchMap[key](copy[key]);
        }
      }
    }

    return copy;
  }

  /**
   * Checks for structural equality:
   *
   *   const book1 = new Book('TypeScript Deep Dive');
   *   const book2 = new Book('TypeScript handbook');
   *
   *   book1.equals(book2); // false
   *   book1.equals(new Book('TypeScript Deep Dive')); // true
   *
   * Note that the comparision is shallow!
   */
  equals(other: this | null | undefined): other is this {
    return (
      other != null &&
      (this === other ||
        Object.getOwnPropertyNames(this).every(p =>
          other[p] instanceof Copyable && this[p] instanceof Copyable
            ? this[p].equals(other[p])
            : other[p] === this[p]
        ))
    );
  }

  /**
   * Returns a string representation of the object that is a bit nicer
   * than the default one (`'[object Object]'`):
   *
   *   const person = new Person('John', 25);
   *   person.toString();
   *   // Person(
   *   //   name="John",
   *   //   age=25
   *   // )
   */
  toString(): string {
    return copyablePrinter(this);
  }
}
