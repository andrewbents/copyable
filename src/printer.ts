import { Copyable } from './copyable';
import { getConstructor } from './utils';

interface ObjectPrinterOptions {
  /**
   * Open parenthesis that is printed before object properties.
   */
  openParen: string;
  /**
   * Close parenthesis that is printed after object properties.
   */
  closeParen: string;
  /**
   * Delimiter for object properties.
   */
  delimiter: string;
  /**
   * Number of spaces to indent object properties with.
   */
  indent: number;
  /**
   * Printed before the value of a property.
   */
  eq: string;
}

/**
 * Print instance of Copyable object with the following format:
 *    Classname(
 *      property1=value1,
 *      property2="value2",
 *      property3=true
 *    )
 */
export const copyablePrinter = objectPrinter({
  openParen: '(\n',
  closeParen: '\n)',
  delimiter: ',\n',
  indent: 2,
  eq: '='
});

/**
 * Print instance of any other object with the following format:
 *    OtherObject { property1: value1, property2: "value2", property3: true }
 */
const objPrinter = objectPrinter({
  openParen: ' { ',
  closeParen: ' }',
  delimiter: ', ',
  indent: 0,
  eq: ': '
});

/**
 * Indents a string with the given number of spaces.
 * Bypasses the string if size=0.
 */
const indentString = (str: string, size = 2) =>
  size === 0 ? str : str.replace(/^(?!\s*$)/gm, ' '.repeat(size));

/**
 * Return a function that prints object instance with the given options.
 */
function objectPrinter(options: ObjectPrinterOptions) {
  return <T extends {}>(obj: T): string => {
    const className = getConstructor(obj).name;
    const properties = indentString(
      printObjectProperties(obj, options),
      options.indent
    );
    return `${className}${options.openParen}${properties}${options.closeParen}`;
  };
}

/**
 * Prints object own properties with the given options.
 */
function printObjectProperties<T extends {}>(
  obj: T,
  { eq, delimiter }: ObjectPrinterOptions
): string {
  const ownPropertyNames = Object.getOwnPropertyNames(obj);
  const last = ownPropertyNames.length - 1;
  return ownPropertyNames.reduce(
    (acc, key, i) =>
      acc + `${key}${eq}${printVal(obj[key])}${i === last ? '' : delimiter}`,
    ``
  );
}

/**
 * Prints a value:
 *   If the value is a string, wraps it with ""
 *   If the value is a Copyable object, delegates it to copyablePrinter
 *   If the value is an object, delegates it to objPrinter
 *   Otherwise the value is converted to a string.
 */
function printVal(val: any): string {
  if (typeof val === 'object' && val != null) {
    return val instanceof Copyable ? copyablePrinter(val) : objPrinter(val);
  }
  return typeof val === 'string' ? `"${val}"` : `${val}`;
}
