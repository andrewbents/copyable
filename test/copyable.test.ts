import { Copyable } from '../src/copyable';

describe('Copyable', () => {
  class Test extends Copyable {
    constructor(readonly num = 0) {
      super();
    }

    testMethod() {
      return 5;
    }
  }

  describe('basic', () => {
    it('should copy', () => {
      const test = new Test(2);

      const copy = test.copy({ num: 5 });
      expect(copy.num).toBe(5);
      expect(test.num).toBe(2);

      expect(() => copy.testMethod()).not.toThrow();
      expect(() => new Test().copy()).not.toThrow();
    });

    it('should copy by key', () => {
      const test1 = new Test(5);
      expect(test1.copy('num', 6).num).toBe(6);
    });

    it('should check equality', () => {
      const test1 = new Test(2);
      const test2 = new Test(2);
      const test3 = new Test(5);

      expect(test1.equals(test1)).toBeTruthy();
      expect(test1.equals(test1.copy())).toBeTruthy();

      expect(test1.equals(test2)).toBeTruthy();
      expect(test1.equals(test3)).not.toBeTruthy();
      expect(test1.equals(null)).not.toBeTruthy();
      expect(test1.equals(undefined)).not.toBeTruthy();
    });

    it('should check equality with nested copyable objects', () => {
      class WithNested extends Copyable {
        constructor(readonly bool: boolean, readonly nested: Test) {
          super();
        }
      }

      const withNested1 = new WithNested(true, new Test(5));
      const withNested2 = new WithNested(true, new Test(5));

      expect(withNested1.equals(withNested2)).toBe(true);

      expect(withNested1.equals(withNested1.copy('bool', false))).toBe(false);
    });

    it('should map', () => {
      const test1 = new Test(7);

      const copy = test1.map({ num: n => n + 2 });
      expect(copy.num).toBe(9);

      expect(test1.map().num).toBe(7);
    });

    it('should map by key', () => {
      const test1 = new Test(7);

      const copy = test1.map('num', num => num + 2);
      expect(copy.num).toBe(9);
    });
  });

  describe('inheritance', () => {
    class Derived extends Test {
      constructor(readonly field: string, num: number) {
        super(num);
      }

      method() {
        return 'method';
      }

      testMethod() {
        return 9;
      }
    }

    const derived = new Derived('test', 5);

    it('should create derived class instance', () => {
      expect(derived.field).toBe('test');
      expect(derived.num).toBe(5);
    });

    it('should copy', () => {
      const copy = derived.copy({ num: -5 });
      expect(copy.num).toBe(-5);
    });

    it('should have correct methods', () => {
      expect(derived.method()).toBe('method');
      expect(derived.testMethod()).toBe(9);
    });
  });

  describe('toString', () => {
    class Inner extends Copyable {
      constructor(readonly bool = true, readonly str = 'hello world') {
        super();
      }
    }

    class Outer extends Copyable {
      constructor(
        readonly num: number = 10,
        readonly innerObj = new Inner(),
        readonly falsy = undefined,
        readonly falsyNull = null,
        readonly plainObj = { a: 5, b: 6 }
      ) {
        super();
      }
    }

    it('should print copyable object', () => {
      expect(new Outer(10, new Inner()).toString()).toMatchSnapshot();
    });

    it('should correctly print subclasses', () => {
      class Derived extends Inner {
        constructor(bool = true, str = 'hello world', readonly derivedVal = 5) {
          super(bool, str);
        }
      }

      expect(new Derived().toString()).toMatchSnapshot();
    });
  });
});
