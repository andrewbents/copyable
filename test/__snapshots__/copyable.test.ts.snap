// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Copyable toString should correctly print subclasses 1`] = `
"Derived(
  bool=true,
  str=\\"hello world\\",
  derivedVal=5
)"
`;

exports[`Copyable toString should print copyable object 1`] = `
"Outer(
  num=10,
  innerObj=Inner(
    bool=true,
    str=\\"hello world\\"
  ),
  falsy=undefined,
  falsyNull=null,
  plainObj=Object { a: 5, b: 6 }
)"
`;
